### DISCUSSION TITLE
Strongswan v5.8.2 is assigning offline leases during client reauthentication

### DISCUSSION THREAD

**Discussion by pk-wc** [strongswan/strongswan#2472] [Q&A]
Hi,

### Scenario:
We are testing with two different clients (say client-1 & client-2), where re-authentication and make_before_break are enabled on both the clients.
These clients are connected to two different Strongswan servers (say server-1 & server-2) and tunnels are coming up successfully on both the clients.
Client versions: 5.9.14
Server versions: 5.8.2

### Issue observed:
During every re-auth with server-1, the TIA on client-1 keeps toggling between two IPs. This is affecting communication of client-1 with backend systems whenever the TIA changes.
While on the other client i.e. client-2 TIA remains same always (as expected).

### Initial Analysis:
On checking the logs at the server-1, we found that server-1 is always assigning the IP with **offline** lease during re-auth.
Whereas, server-2 is always assigning the IP with **online** lease during re-auth.

### Expectation:
During re-auth, TIA is not supposed to change on client-1.
Can you please help us identify the reason for TIA toggle on client-1 ?

### Configurations:
Below are the configurations of strongswan server:
**Server-1 (toggling TIA):**
File: ipsec.conf
```
config setup
        uniqueids = never
        charondebug="ike 4, chd 1, cfg 2, net 1, enc 1, lib 1, mgr 1, knl 1, dmn 4"
 
conn home
        left=172.29.17.155
        leftid=<masked>
        leftsubnet=::0/0
        leftcert=<masked>
        right=%any
        rightid=%any
        rightcert=%any
        ike=aes256-sha1-prfsha1-modp2048!
        esp=aes128-sha1,aes256-sha1
        mobike=no
        ikelifetime=90000s
        lifetime=90000s
        auto=add
        reauth=no
        rekeymargin=3m
        keyingtries=1
        keyexchange=ikev2
        authby=pubkey
        dpdaction=clear
        rightdns=<masked>
        rightsourceip=fd00:0005:B954::/112
        compress=yes
```

File: strongswan.conf
```
charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
}
include strongswan.d/*.conf
```

**Server-2 (Constant TIA):**
File: ipsec.conf
```
config setup
    uniqueids = never
    charondebug="ike 4, chd 1, cfg 1, net 1, enc 1, lib 1, mgr 1, knl 1 dmn 1"
  
conn %default
        ikelifetime=86400s
        keylife=36000s
        rekeymargin=1m
        keyingtries=1
        keyexchange=ikev2
        ike=aes192-sha256-prfsha256-modp3072,aes256-sha1-modp1024,aes-sha256-prfsha256-modp2048,aes256-sha1-modp1024,aes256-sha1-modp2048,aes-prfsha256-sha256-modp2048
        esp=aes192-sha384,aes256-sha1,aes-sha1,aes256-sha256,aes256-sha256-modp2048
        mobike=no
        auto=add
        reauth=no
        dpdaction=clear
        fragmentation=yes
 
conn rw
        left=%any
        leftcert=<masked>
        leftid=%any
        leftsubnet=0.0.0.0/0,::0/0
        rightsourceip=2008:1:161:25::/64,10.161.25.0/24
        right=%any
        rightid=%any
        rightcert=%any
        authby=pubkey
        auto=add
        rightdns=<masked>
```

File: strongswan.conf
```
charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
        cert_cache = no
}
include strongswan.d/*.conf
```

Please let me know if additional logs/configurations are needed.

Thanks,
Prasad

**Comment by tobiasbrunner** [comment]
Please provide full server logs (and maybe client logs as well), see #196 for log levels.

**Reply by pk-wc** [reply]
[charon_logs.txt](https://github.com/user-attachments/files/17167415/charon_logs.txt)
Attached are the Server-1's charon logs. Look for the client with ID 'clientserialno.someorg.com'.

We suspected that the existing IP leases could be the reason for assigning IP with an offline lease. So, we performed "ipsec restart" on Server-1 and it started working fine. But, "ipsec restart" was a costly operation for us as it impacted all the other clients that were connected to Server-1.
So, if existing IP leases is the RCA for this issue then is there a way we can clear existing leases for a specific client ? If so, can you please share the corresponding command ?

**Reply by tobiasbrunner** [reply]
Since it's not the full log, we don't have any information on the IKE_SA (or IKE_SAs if that identity was used multiple times) before the reauthentication. Maybe there were concurrent IKE_SAs established from different IPs/ports, so multiple leases were assigned to the identity.

If that was the case and one of the SAs was closed and its lease set offline, it will actually get assigned first during a make-before-break reauthentication. Looking at the code, I see that the lookup for offline leases happens before checking for an online lease. That's obviously not ideal when using make-before-break reauthentications combined with having uniqueness checks disabled i.e. allowing multiple connections for the same identity from different endpoints.

We should probably reverse this, see the possible fix in the _2472-pool-existing-reverse_ branch. I don't think it should have any negative side-effects as we only assign online leases to clients that connect from the same exact IP and port.

**Reply by tobiasbrunner** [reply]
This fix is now in master (d4a0dd9f93d67648eed1ebb223c6cb853101a80a).
